core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -1136,9 +1128,59 @@ private boolean shouldPerformTransactionOperation(StoreStatus storeStatus, Predi
       ).toCompletionStage();
    }
 
-   private <K, V> MVCCEntry<K, V> acquireKeyFromContext(InvocationContext ctx, WriteCommand command, Object key,
-         BiPredicate<WriteCommand, Object> commandKeyPredicate) {
-      if (commandKeyPredicate.test(command, key)) {
+   private <I> Flowable<NonBlockingStore.SegmentedPublisher<I>> filterSharedSegments(
+         Flowable<NonBlockingStore.SegmentedPublisher<I>> flowable, AtomicLong writeCount) {
+      return flowable.filter(sp -> {
+         if (distributionManager.getCacheTopology().getSegmentDistribution(sp.getSegment()).isPrimary()) {
+            return true;
+         }
+         // Way to ensure the filtered groups are still subscribed and the values are not counted
+         Flowable.fromPublisher(sp)
+               .count()
+               .subscribe(ignoredCount -> {
+                  if (writeCount != null) {
+                     writeCount.addAndGet(-ignoredCount);
+                  }
+               });
```

Comment:
Is there a better way of doing this? I don't have an alternative suggestion but this smells funny.